File: generate-posts.mjs
#!/usr/bin/env node

/**
 * Simple Blog Post Generator for RAIX
 * Converts Markdown files to posts.json
 * 
 * Usage: node generate-posts.mjs
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

console.log('\n🚀 RAIX Blog Post Generator\n');

// Parse frontmatter
function parseFrontmatter(content) {
  const match = content.match(/^---\s*\n([\s\S]*?)\n---\s*\n([\s\S]*)$/);
  if (!match) return { data: {}, content };
  
  const data = {};
  match[1].split('\n').forEach(line => {
    const idx = line.indexOf(':');
    if (idx > 0) {
      data[line.substring(0, idx).trim()] = line.substring(idx + 1).trim();
    }
  });
  
  return { data, content: match[2] };
}

// Convert markdown to HTML
function md2html(md) {
  return md
    .replace(/^### (.*$)/gim, '<h3>$1</h3>')
    .replace(/^## (.*$)/gim, '<h2>$1</h2>')
    .replace(/^# (.*$)/gim, '<h1>$1</h1>')
    .replace(/\*\*(.+?)\*\*/g, '<strong>$1</strong>')
    .replace(/\*(.+?)\*/g, '<em>$1</em>')
    .replace(/\[([^\]]+)\]\(([^)]+)\)/g, '<a href="$2">$1</a>')
    .replace(/`([^`]+)`/g, '<code>$1</code>')
    .split('\n')
    .map(line => {
      if (line.trim().startsWith('- ')) return '<li>' + line.trim().substring(2) + '</li>';
      if (line.trim() && !line.trim().startsWith('<')) return '<p>' + line + '</p>';
      return line;
    })
    .join('\n')
    .replace(/(<li>.*<\/li>\n)+/g, match => '<ul>' + match + '</ul>');
}

// Main
const postsDir = path.join(__dirname, 'posts');
const outputFile = path.join(__dirname, 'assets', 'posts.json');

if (!fs.existsSync(postsDir)) {
  console.error('❌ posts/ folder not found!\n');
  process.exit(1);
}

const files = fs.readdirSync(postsDir).filter(f => f.endsWith('.md') && !f.startsWith('_'));

if (files.length === 0) {
  console.error('❌ No .md files found in posts/\n');
  process.exit(1);
}

console.log(`📝 Found ${files.length} post(s):\n`);

const posts = files.map(file => {
  const content = fs.readFileSync(path.join(postsDir, file), 'utf8');
  const { data, content: md } = parseFrontmatter(content);
  const slug = file.replace('.md', '');
  
  console.log(`   ✓ ${file}`);
  
  return {
    id: slug,
    slug,
    title: data.title || 'Untitled',
    excerpt: data.excerpt || '',
    content: md2html(md),
    author: data.author || 'Roberto',
    date: data.date || new Date().toISOString().split('T')[0],
    category: data.category || 'General',
    image: data.image || '/hero-image.png',
    readTime: Math.ceil(md.split(/\s+/).length / 200) + ' min read'
  };
}).sort((a, b) => new Date(b.date) - new Date(a.date));

fs.writeFileSync(outputFile, JSON.stringify(posts, null, 2));

console.log(`\n✅ Generated: ${outputFile}`);
console.log(`📊 Total posts: ${posts.length}\n`);
console.log('🚀 Ready to deploy!\n');

File: generate-posts.py
#!/usr/bin/env python3
"""
RAIX Blog Post Generator
Converts Markdown files to posts.json

Usage: python3 generate-posts.py
"""

import os
import json
import re
from datetime import datetime

print('\n🚀 RAIX Blog Post Generator\n')

def parse_frontmatter(content):
    """Extract frontmatter from markdown"""
    match = re.match(r'^---\s*\n(.*?)\n---\s*\n(.*)$', content, re.DOTALL)
    if not match:
        return {}, content
    
    data = {}
    for line in match.group(1).split('\n'):
        if ':' in line:
            key, value = line.split(':', 1)
            data[key.strip()] = value.strip()
    
    return data, match.group(2)

def md_to_html(md):
    """Simple markdown to HTML conversion"""
    html = md
    html = re.sub(r'^### (.*)$', r'<h3>\1</h3>', html, flags=re.MULTILINE)
    html = re.sub(r'^## (.*)$', r'<h2>\1</h2>', html, flags=re.MULTILINE)
    html = re.sub(r'^# (.*)$', r'<h1>\1</h1>', html, flags=re.MULTILINE)
    html = re.sub(r'\*\*(.+?)\*\*', r'<strong>\1</strong>', html)
    html = re.sub(r'\*(.+?)\*', r'<em>\1</em>', html)
    html = re.sub(r'\[([^\]]+)\]\(([^)]+)\)', r'<a href="\2">\1</a>', html)
    html = re.sub(r'`([^`]+)`', r'<code>\1</code>', html)
    
    lines = []
    for line in html.split('\n'):
        line = line.strip()
        if line.startswith('- '):
            lines.append('<li>' + line[2:] + '</li>')
        elif line and not line.startswith('<'):
            lines.append('<p>' + line + '</p>')
        else:
            lines.append(line)
    
    html = '\n'.join(lines)
    html = re.sub(r'(<li>.*</li>\n)+', lambda m: '<ul>' + m.group(0) + '</ul>', html)
    
    return html

# Main
posts_dir = 'posts'
output_file = 'assets/posts.json'

if not os.path.exists(posts_dir):
    print('❌ posts/ folder not found!\n')
    exit(1)

files = [f for f in os.listdir(posts_dir) if f.endswith('.md') and not f.startswith('_')]

if not files:
    print('❌ No .md files found in posts/\n')
    exit(1)

print(f'📝 Found {len(files)} post(s):\n')

posts = []
for filename in files:
    with open(os.path.join(posts_dir, filename), 'r', encoding='utf-8') as f:
        content = f.read()
    
    data, md = parse_frontmatter(content)
    slug = filename.replace('.md', '')
    
    print(f'   ✓ {filename}')
    
    posts.append({
        'id': slug,
        'slug': slug,
        'title': data.get('title', 'Untitled'),
        'excerpt': data.get('excerpt', ''),
        'content': md_to_html(md),
        'author': data.get('author', 'Roberto'),
        'date': data.get('date', datetime.now().strftime('%Y-%m-%d')),
        'category': data.get('category', 'General'),
        'image': data.get('image', '/hero-image.png'),
        'readTime': f"{max(1, len(md.split()) // 200)} min read"
    })

posts.sort(key=lambda x: x['date'], reverse=True)

with open(output_file, 'w', encoding='utf-8') as f:
    json.dump(posts, f, indent=2, ensure_ascii=False)

print(f'\n✅ Generated: {output_file}')
print(f'📊 Total posts: {len(posts)}\n')
print('🚀 Ready to deploy!\n')

